### PR TITLE
first pass at getting logits to the client code

### PR DESCRIPTION
**Pull request by ftyers**
First pass at maintaining the output of the acoustic model and returning it to the client. Example output is [here](https://dpaste.com/F4BSLA3NC). Comments welcome.

### CODE REVIEW

File: native_client/coqui-stt.h [hunk 1]
```diff
@@ -61,6 +61,11 @@ typedef struct Metadata {
   const CandidateTranscript* const transcripts;
   /** Size of the transcripts array */
   const unsigned int num_transcripts;
+  /** Logits and information to decode them **/
+  const char* const alphabet;
```
Comment (reuben) [inline native_client/coqui-stt.h:65]: I guess this is just for the first pass but the output data structures should have decoded characters rather than including an Alphabet serialization in it.

File: native_client/ctcdecode/ctc_beam_search_decoder.cpp [hunk 2]
```diff
@@ -57,6 +57,12 @@ DecoderState::init(const Alphabet& alphabet,
   return 0;
 }
 
+void
+DecoderState::withLogits(bool val)
+{
+  keep_logits_ = val;
+}
+
```
Comment (reuben) [inline native_client/ctcdecode/ctc_beam_search_decoder.cpp:65]: We should set this in `::init`, we don't want to have to reason about already initialized `DecoderState`s which then enable saving of logits.

File: native_client/ctcdecode/ctc_beam_search_decoder.cpp [hunk 3]
```diff
@@ -93,6 +99,15 @@ DecoderState::next(const double *probs,
       full_beam = (num_prefixes == beam_size_);
     }
 
+    // save the softmax of the current timestep
+    if(keep_logits_) {
```
Comment (reuben) [inline native_client/ctcdecode/ctc_beam_search_decoder.cpp:103]: nit: space after `if`, `else`, `for`, `while`, `do`

File: native_client/ctcdecode/ctc_beam_search_decoder.h [hunk 4]
```diff
@@ -45,6 +47,8 @@ class DecoderState
    *     ext_scorer: External scorer to evaluate a prefix, which consists of
    *                 n-gram language model scoring and word insertion term.
    *                 Default null, decoding the input sample without scorer.
+   *     hot_words: The hotwords and their scores
+   *     keep_logits: Should we return the output of the AM?
```
Comment (reuben) [inline native_client/ctcdecode/ctc_beam_search_decoder.h:51]: This parameter doesn't exist, but it should, instead of `withLogits`.

File: native_client/stt.cc [hunk 5]
```diff
@@ -401,6 +403,8 @@ int STT_SetScorerAlphaBeta(ModelState* aCtx,
   return STT_ERR_SCORER_NOT_ENABLED;
 }
 
+
+
```
Comment (reuben) [inline native_client/stt.cc:407]: nit: unneeded whitespace change

File: native_client/stt.cc [hunk 6]
```diff
@@ -431,6 +435,7 @@ STT_CreateStream(ModelState* aCtx,
                            aCtx->scorer_,
                            aCtx->hot_words_);
 
+
```
Comment (reuben) [inline native_client/stt.cc:438]: nit: unneeded whitespace change

**Review by reuben — COMMENTED** [review]
Did a first pass with some comments!

File: native_client/coqui-stt.h (hunk 1 above)
Comment (ftyers) [reply native_client/coqui-stt.h:65]: Yeah, it's the first pass. I didn't want to do anything fancy because I'd probably have to change it anway. What is the suggestion for a C-compatible structure for encoding a matrix of double X character? Should I make some structs, something like:

```c
struct Symbol {
    char *symbol;
}
struct Timestep {
    Symbols **symbols;
    double **softmax;
} 
```
Then just include `Timestep **timesteps` ?

The issue is that because of UTF-8 stuff, we can't assume that `char*` is a single symbol. So we have to wrap it in some more structure.

File: native_client/ctcdecode/ctc_beam_search_decoder.cpp (hunk 2 above)
Comment (ftyers) [reply native_client/ctcdecode/ctc_beam_search_decoder.cpp:65]: That's fine, better even I just didn't want to add another parameter without checking first. I'll change that now.

File: native_client/ctcdecode/ctc_beam_search_decoder.cpp (hunk 3 above)
Comment (ftyers) [reply native_client/ctcdecode/ctc_beam_search_decoder.cpp:103]: I'll change this one, do you want me to change others that I find?

Comment (reuben) [reply native_client/ctcdecode/ctc_beam_search_decoder.cpp:103]: in your code, yes. in existing code, no.

File: native_client/ctcdecode/ctc_beam_search_decoder.cpp (hunk 2 above)
Comment (ftyers) [reply native_client/ctcdecode/ctc_beam_search_decoder.cpp:65]: Ah, now I remember why there was a problem here...

The decoder is initialised in

```c
STT_CreateStream(ModelState* aCtx,
                StreamingState** retval)
```

I didn't want to add another parameter to an API function, or proliferate new API calls. 

A couple of options are:
1. Add a `STT_CreateStreamWithLogits()` API call.
2. Change the `STT_CreateStream()` to include a `bool keep_logits` parameter.
3. Add `keep_logits_` variable to `ModelState` (we had this before, but I include it for completeness)

Any options I've missed?

File: native_client/coqui-stt.h (hunk 1 above)
Comment (reuben) [reply native_client/coqui-stt.h:65]: C makes this type of stuff quite verbose. I recommend using a linear array (double*, char*) and having the array dimensions in separate attributes (size of time dimension and size of class dimension), and then document if the memory layout is time major or class major. We also need to encode the CTC blank character as a char* somehow (using a special ASCII symbol maybe?).

So maybe something like:

```cpp
struct AcousticModelEmissions {
    int num_symbols;
    char** symbols; // num_symbols long array of NULL-terminated strings
    int num_timesteps;
    double** emissions; // num_timesteps long array, each pointer is a num_symbols long array
}

struct Metadata {
    // ...
    AcousticModelEmissions *emissions;
}
```

Comment (reuben) [reply native_client/coqui-stt.h:65]: (This would be a time major layout, you could also encode `emissions` in a symbol major layout. I guess time major is easier to implement as the decoder is already operating in a time major way.

File: native_client/ctcdecode/ctc_beam_search_decoder.cpp (hunk 2 above)
Comment (reuben) [reply native_client/ctcdecode/ctc_beam_search_decoder.cpp:65]: I'd go with:

> 1. Add a STT_CreateStreamWithLogits() API call.

But not an API call, just a private abstraction. See for example `CreateStreamAndFeedAudioContent`.

File: native_client/coqui-stt.h (hunk 1 above)
Comment (ftyers) [reply native_client/coqui-stt.h]: Ok, implemented. I made a minor change, which was to use `double* emissions` as I was doing before. I can change it, but could do with a pointer to how to do it (pun intended).

**Comment by reuben** [comment]
Also, please sign the CLA!

File: native_client/modelstate.cc [hunk 7]
```diff
@@ -65,11 +65,46 @@ ModelState::decode_metadata(const DecoderState& state,
     memcpy(&transcripts[i], &transcript, sizeof(CandidateTranscript));
   }
 
+  unsigned int num_timesteps = out[0].logits.size();
+  unsigned int alphabet_size = alphabet_.GetSize();
```
[the remark this answers is not in the document]
Comment (ftyers) [reply native_client/modelstate.cc:69]: I'm happy to add in the blank symbol, is it just a matter of doing `+ 1` to `alphabet_size` ?

File: native_client/alphabet.h [hunk 8]
```diff
@@ -28,6 +28,9 @@ class Alphabet : public fl::lib::text::Dictionary
   // Serialize alphabet into a binary buffer.
   std::string Serialize();
 
+  // Serialize alphabet into a compact text representation
+  std::string SerializeLine();
+
```
[the remark this answers is not in the document]
Comment (ftyers) [reply native_client/alphabet.h]: Fixed.

File: native_client/alphabet.cc [hunk 9]
```diff
@@ -80,6 +80,19 @@ Alphabet::InitFromLabels(const std::vector<std::string>& labels)
...
+{
+  std::stringstream out;
+
+  for (int idx = 0; idx < entrySize(); ++idx) {
+    out << getEntry(idx) << "\n";
+  }
+
+  return out.str();
+}
+
```
[the remark this answers is not in the document]
Comment (ftyers) [reply native_client/alphabet.cc]: Fixed.

File: native_client/coqui-stt.h [hunk 10]
```diff
@@ -61,6 +73,8 @@ typedef struct Metadata {
   const CandidateTranscript* const transcripts;
   /** Size of the transcripts array */
   const unsigned int num_transcripts;
+  /** Logits and information to decode them **/
+  AcousticModelEmissions *emissions;
```
[the remark this answers is not in the document]
Comment (ftyers) [reply native_client/coqui-stt.h]: Done.

File: native_client/modelstate.cc [hunk 11]
```diff
@@ -65,11 +65,46 @@ ModelState::decode_metadata(const DecoderState& state,
...
   Metadata* ret = (Metadata*)malloc(sizeof(Metadata));
-  Metadata metadata {
-    transcripts,  // transcripts
-    num_returned, // num_transcripts
-  };
-  memcpy(ret, &metadata, sizeof(Metadata));
-  return ret;
+
+  if (num_timesteps > 0) { // see if the logit structure has been filled and returned
+
```
[the remark this answers is not in the document]
Comment (ftyers) [reply native_client/modelstate.cc]: Removed.

File: native_client/modelstate.cc [hunk 12]
```diff
@@ -65,11 +65,46 @@ ModelState::decode_metadata(const DecoderState& state,
...
+    AcousticModelEmissions* emissions = (AcousticModelEmissions*)malloc(sizeof(AcousticModelEmissions));
+
+    emissions->num_symbols = alphabet_size;
+    emissions->num_timesteps = num_timesteps;
+    emissions->symbols = (char**)malloc(sizeof(char*)*alphabet_size);
+    for(int i = 0; i < alphabet_size; i++) {
+        emissions->symbols[i] = strdup(alphabet_.DecodeSingle(i).c_str());
+    }
+
+    double* logits = (double*)malloc(sizeof(double)*alphabet_size*num_timesteps);
```
[the remark this answers is not in the document]
Comment (ftyers) [reply native_client/modelstate.cc]: I've changed it in this place, but should I change it across the board (e.g. in `struct Output` too)?

File: native_client/modelstate.cc [hunk 13]
```diff
@@ -65,11 +65,46 @@ ModelState::decode_metadata(const DecoderState& state,
...
-  return ret;
+
+  if (num_timesteps > 0) { // see if the logit structure has been filled and returned
+
+    AcousticModelEmissions* emissions = (AcousticModelEmissions*)malloc(sizeof(AcousticModelEmissions));
+
+    emissions->num_symbols = alphabet_size;
+    emissions->num_timesteps = num_timesteps;
+    emissions->symbols = (char**)malloc(sizeof(char*)*alphabet_size);
+    for(int i = 0; i < alphabet_size; i++) {
```
[the remark this answers is not in the document]
Comment (ftyers) [reply native_client/modelstate.cc]: Fixed.

File: native_client/coqui-stt.h [hunk 14]
```diff
@@ -53,6 +53,18 @@ typedef struct CandidateTranscript {
   const double confidence;
 } CandidateTranscript;
 
+/**
+ * @brief An structure to contain emissions (the softmax output of individual
+ *        timesteps) from the acoustic model.
+ *        The layout is time major.
```
[the remark this answers is not in the document]
Comment (ftyers) [reply native_client/coqui-stt.h]: I've specified the layout. I used `@member` as I couldn't find any equivalent in the other comments/docstrings for `structs`, and I thought that `@param` doesn't sound quite right.

File: native_client/modelstate.cc (hunk 7 above)
[the remark this answers is not in the document]
Comment (reuben) [reply native_client/modelstate.cc:69]: And handling the representation as a char* later.

[the remark this answers is not in the document]
Comment (ftyers) [reply native_client/modelstate.cc:69]: So, something like:
- add 1, 
- have a char representation, something like `@0@` or `@EPS@`  ?
Any preferences for the character representation?

File: native_client/modelstate.cc (hunk 12 above)
[the remark this answers is not in the document]
Comment (reuben) [reply native_client/modelstate.cc]: Only in code you're adding.

[the remark this answers is not in the document]
Comment (ftyers) [reply native_client/modelstate.cc]: I think I added it in `struct Output`, and do you want it in the arguments / variables too? e.g. `keep_probs` instead of `keep_logits` and `WithProbs` instead of `WithLogits` ?

File: native_client/modelstate.cc (hunk 7 above)
[the remark this answers is not in the document]
Comment (reuben) [reply native_client/modelstate.cc:69]: Or maybe a single tab character? 0x9? I can't think of a reason why anyone wouldn't normalize whitespaces in the data so it seems like an encoding that won't conflict.

File: native_client/coqui-stt.h (hunk 14 above)
[the remark this answers is not in the document]
Comment (reuben) [reply native_client/coqui-stt.h]: It's just whatever `/** */` comment is immediately before the attribute.

File: native_client/modelstate.cc (hunk 7 above)
[the remark this answers is not in the document]
Comment (reuben) [reply native_client/modelstate.cc:69]: Any reserved ASCII token for machine communication should do the job. Though many aren't visible.

File: native_client/modelstate.cc (hunk 12 above)
[the remark this answers is not in the document]
Comment (reuben) [reply native_client/modelstate.cc]: Better yet `WithEmissions` to match the struct returned.

File: native_client/modelstate.cc (hunk 7 above)
[the remark this answers is not in the document]
Comment (ftyers) [reply native_client/modelstate.cc:69]: Ok, this is done now.

File: native_client/coqui-stt.h (hunk 14 above)
[the remark this answers is not in the document]
Comment (ftyers) [reply native_client/coqui-stt.h]: Done.

File: native_client/coqui-stt.h [hunk 15]
```diff
@@ -53,6 +53,18 @@ typedef struct CandidateTranscript {
...
+ */
+typedef struct AcousticModelEmissions {
+  int num_symbols;
+  char** symbols; // num_symbols long array of NUL-terminated strings
+  int num_timesteps;
+  double* emissions; // num_timesteps long array, each pointer is a num_symbols long array
```
[the remark this answers is not in the document]
Comment (ftyers) [reply native_client/coqui-stt.h]: Done.

File: native_client/modelstate.cc (hunk 12 above)
[the remark this answers is not in the document]
Comment (ftyers) [reply native_client/modelstate.cc]: Ok done.

**Comment by ftyers** [comment]
Comments from Matrix:
imatge

**Comment by ftyers** [comment]
@reuben so in this case I should collect them in `processBatch`, store them in `StreamingState`? The problem there is that the Metadata is returned in:
```
Metadata*
ModelState::decode_metadata(const DecoderState& state,
                            size_t num_results)
```
which has access to only `ModelState` and `DecoderState`. 

So then I would edit:
```
Metadata*
StreamingState::intermediateDecodeWithMetadata(unsigned int num_results) const
```

And have it fill in the logits there?

**Comment by ftyers** [comment]
Ok, I moved it to `StreamingState`. There is something I don't like, which is having to reallocate the memory for the metadata because we can't just update the struct apparently:
```
native_client/stt.cc: In member function 'Metadata* StreamingState::finishStreamWithMetadata(unsigned int)':
native_client/stt.cc:175:18: error: assignment of read-only member 'Metadata::emissions'
  175 |     m->emissions = emissions;
      |     ~~~~~~~~~~~~~^~~~~~~~~~~
```
Comments welcome!

**Comment by ftyers** [comment]
Someone else will work on this who can sign the CLA.
